x-pack/plugins/dataset_quality/public/hooks/use_dataset_quality_table.tsx

```diff
@@ -28,63 +28,83 @@ export const useDatasetQualityTable = () => {
+  const loading = useSelector(service, (state) => state.matches('datasets.fetching'));
+
+  const degradedStats = useSelector(service, (state) => state.context.degradedDocStats);
+  const loadingDegradedStats = useSelector(service, (state) =>
+    state.matches('degradedDocs.fetching')
+  );
+
+  const flyoutClosed = useSelector(service, (state) => state.matches('datasets.loaded.idle'));
+
+  const closeFlyout = useCallback(() => service.send({ type: 'CLOSE_FLYOUT' }), [service]);
+  const openFlyout = useCallback(
+    (selectedDataset: DataStreamStat) => {
+      if (flyout?.dataset?.rawName === selectedDataset.rawName) {
+        service.send({
+          type: 'CLOSE_FLYOUT',
+        });
+
+        return;
+      }
+
+      service.send({
+        type: flyoutClosed ? 'OPEN_FLYOUT' : 'SELECT_NEW_DATASET',
+        dataset: selectedDataset as FlyoutDataset,
+      });
+    },
+    [flyout?.dataset?.rawName, flyoutClosed, service]
   );
 
   const columns = useMemo(
     () =>
       getDatasetQualityTableColumns({
         fieldFormats,
-        selectedDataset,
-        setSelectedDataset,
+        selectedDataset: flyout?.dataset,
+        openFlyout,
         loadingDegradedStats,
       }),
-    [fieldFormats, loadingDegradedStats, selectedDataset, setSelectedDataset]
+    [flyout?.dataset, fieldFormats, loadingDegradedStats, openFlyout]
   );
 
   const pagination = {
-    pageIndex,
-    pageSize,
-    totalItemCount: data.length,
+    pageIndex: page,
+    pageSize: rowsPerPage,
+    totalItemCount: dataStreamStats.length,
     hidePerPageOptions: true,
   };
 
   const onTableChange = useCallback(
     (options: {
       page: { index: number; size: number };
-      sort?: { field: SORT_FIELD; direction: DIRECTION };
+      sort?: { field: SortField; direction: Direction };
     }) => {
-      setPageIndex(options.page.index);
-      setPageSize(options.page.size);
-      setSortField(options.sort?.field || DEFAULT_SORT_FIELD);
-      setSortDirection(options.sort?.direction || DEFAULT_SORT_DIRECTION);
+      service.send({
+        type: 'UPDATE_TABLE_CRITERIA',
+        criteria: {
+          page: options.page.index,
+          rowsPerPage: options.page.size,
+          sort: {
+            field: options.sort?.field || DEFAULT_SORT_FIELD,
+            direction: options.sort?.direction || DEFAULT_SORT_DIRECTION,
+          },
+        },
+      });
     },
-    []
+    [service]
   );
 
-  const sort = {
-    sort: { field: sortField, direction: sortDirection },
-  };
-
   const renderedItems = useMemo(() => {
-    const overridenSortingField = sortingOverrides[sortField] || sortField;
-    const mergedData = data.map((dataStream) => {
+    const overridenSortingField = sortingOverrides[sort.field] || sort.field;
+    const mergedData = dataStreamStats.map((dataStream) => {
       const degradedDocs = find(degradedStats, { dataset: dataStream.rawName });
```

Comment:
**suggestion**: I remember we had an offline conversation about having this merge happen when the state machine resolves the degraded docs, so this logic happens in the machine flow when the loaded state occurs.

Reply:
The thing I didn't like with that approach is we would need to disable `predictableActionArguments: true` which from what I read makes the machine compatible with version 5 (in case we want to upgrade, which I also saw as a potential task in onWeek)

Reply:
Could you explain why it would require disabling the `predictableActionArguments` option? As far as I see, we just need to run this merging operation as an action when the `loadDegradedDocs`service succeeds.

Reply:
My idea was to capture the value of the machine and perform the merge only if both calls have finished. For doing this I would need the argument `meta` that is only available when having `predictableActionArguments = false`

Reply:
You could create an action that performs the merge in both cases when the parallel API requests complete:
- datasets.loaded.actions ➡️ ['storeDatasets', 'storeDatasetsWithDegradedDetail']
- degradedDocs.loaded.actions ➡️ ['storeDegradedDocs, 'storeDatasetsWithDegradedDetail']

In both cases, the data source will be stored, and this action will be executed with a guard that specify both datasets and degradedDocs should exist in context.

Reply:
What happens when you retrigger the API? it could happen that you have an old state of one of the two and the guard is bypassed?

Reply:
If a refresh/reload operation is triggered, what do we expect? IMO, we should refetch the datasets and the degraded docs if the refreshed is specified for the whole page results.
Say instead that we only need to reload the degradedDocs data in background to provide constant feedback, in that case we only need to reload the degradedDocs, while the datasets remain the same.

It depends on what is the defined reload touchpoint, which I understand in this case is the first case.

Reply:
Addressed in [8f14260](https://github.com/elastic/kibana/pull/174906/commits/8f142600f3255437b6c085391a8c2a96596f75c6).